# Rules Synchronization Setup Guide

This guide explains how to set up and use the centralized rules and skills synchronization system across all AI agent platforms.

## Overview

The `.agents/rules/sync-rules.sh` script synchronizes:

- **Rules** - Project guidelines and standards (`.agents/rules/*.md`)
- **Skills** - Agent capabilities and extensions (`.agents/skills/`)

**Source of Truth:** `.agents/` directory
**Synchronized to:** Cursor, Claude Code, Gemini CLI, Antigravity

## Architecture

### Synchronization Strategy

| Platform    | Rules | Skills | Method                           |
| ----------- | ----- | ------ | -------------------------------- |
| Cursor      | ‚úÖ    | ‚úÖ     | Full directory symlinks          |
| Claude Code | ‚úÖ    | ‚úÖ     | Full directory symlinks          |
| Gemini CLI  | ‚úÖ    | ‚úÖ     | Full directory symlinks          |
| Antigravity | ‚úÖ    | ‚úÖ     | Copy (rules), Selective (skills) |

### Directory Structure

```
.agents/
‚îú‚îÄ‚îÄ rules/                  # Source of truth for rules
‚îÇ   ‚îú‚îÄ‚îÄ core-principles.md
‚îÇ   ‚îú‚îÄ‚îÄ code-style.md
‚îÇ   ‚îú‚îÄ‚îÄ documentation.md
‚îÇ   ‚îú‚îÄ‚îÄ git-workflow.md
‚îÇ   ‚îú‚îÄ‚îÄ testing.md
‚îÇ   ‚îú‚îÄ‚îÄ use-context7.md
‚îÇ   ‚îî‚îÄ‚îÄ sync-rules.sh      # Synchronization script
‚îî‚îÄ‚îÄ skills/                 # Source of truth for skills
    ‚îú‚îÄ‚îÄ agent-development/
    ‚îú‚îÄ‚îÄ command-development/
    ‚îú‚îÄ‚îÄ find-skills/
    ‚îú‚îÄ‚îÄ hook-development/
    ‚îú‚îÄ‚îÄ mcp-integration/
    ‚îú‚îÄ‚îÄ skill-creator/
    ‚îî‚îÄ‚îÄ skill-development/

# After sync:
.cursor/
‚îú‚îÄ‚îÄ rules ‚Üí ../.agents/rules    # Symlink
‚îî‚îÄ‚îÄ skills ‚Üí ../.agents/skills  # Symlink

.claude/
‚îú‚îÄ‚îÄ rules ‚Üí ../.agents/rules    # Symlink
‚îî‚îÄ‚îÄ skills ‚Üí ../.agents/skills  # Symlink

.gemini/
‚îú‚îÄ‚îÄ rules ‚Üí ../.agents/rules    # Symlink
‚îî‚îÄ‚îÄ skills ‚Üí ../.agents/skills  # Symlink

.agent/
‚îú‚îÄ‚îÄ rules/                      # Copied files
‚îÇ   ‚îú‚îÄ‚îÄ core-principles.md
‚îÇ   ‚îú‚îÄ‚îÄ code-style.md
‚îÇ   ‚îî‚îÄ‚îÄ ...
‚îî‚îÄ‚îÄ skills/                     # Selective symlinks
```

## Installation

### Prerequisites

- macOS or Linux (symlinks required)
- Bash shell
- Project cloned with Git

### Initial Setup

1. **Verify source directories exist:**

   ```bash
   ls -la .agents/rules
   ls -la .agents/skills
   ```

2. **Make script executable (if not already):**

   ```bash
   chmod +x .agents/rules/sync-rules.sh
   ```

3. **Test with dry-run:**

   ```bash
   ./.agents/rules/sync-rules.sh --dry-run
   ```

4. **Run actual sync:**

   ```bash
   ./.agents/rules/sync-rules.sh
   ```

5. **Verify symlinks created:**
   ```bash
   ls -la .cursor/rules .cursor/skills
   ls -la .claude/rules .claude/skills
   ls -la .gemini/rules .gemini/skills
   ```

## Usage

### Running the Sync Script

**Basic usage:**

```bash
./.agents/rules/sync-rules.sh
```

**Dry-run mode (preview changes):**

```bash
./.agents/rules/sync-rules.sh --dry-run
```

**Expected output:**

```
üîÑ Synchronizing rules and skills from .agents/ to all agent directories...

üìã Validating source directories...
  ‚úÖ Rules source: /path/to/project/.agents/rules
  ‚úÖ Skills source: /path/to/project/.agents/skills

üéØ Syncing Cursor...
  ‚úÖ Created rules symlink: .cursor/rules ‚Üí ../.agents/rules
  ‚úÖ Created skills symlink: .cursor/skills ‚Üí ../.agents/skills

ü§ñ Syncing Claude Code...
  ‚úÖ Created rules symlink: .claude/rules ‚Üí ../.agents/rules
  ‚úÖ Created skills symlink: .claude/skills ‚Üí ../.agents/skills

üíé Syncing Gemini CLI...
  ‚úÖ Created rules symlink: .gemini/rules ‚Üí ../.agents/rules
  ‚úÖ Skills already synced (existing symlink)

üåå Syncing Antigravity (special case)...
  üìù Copying rules to .agent/rules/...
  ‚úÖ Rules copied to .agent/rules/
  ‚úÖ Skills directory exists (selective approach maintained)

üîç Verifying symlinks...
  ‚úÖ Cursor rules: .cursor/rules ‚Üí ../.agents/rules
  ‚úÖ Cursor skills: .cursor/skills ‚Üí ../.agents/skills
  ‚úÖ Claude rules: .claude/rules ‚Üí ../.agents/rules
  ‚úÖ Claude skills: .claude/skills ‚Üí ../.agents/skills
  ‚úÖ Gemini rules: .gemini/rules ‚Üí ../.agents/rules
  ‚úÖ Gemini skills: .gemini/skills ‚Üí ../.agents/skills

‚úÖ Synchronization completed successfully
```

### Adding New Rules

1. **Create rule file in source:**

   ```bash
   # Create new rule
   cat > .agents/rules/security.md << 'EOF'
   # Security Guidelines

   ## Best Practices
   - Never commit secrets
   - Use environment variables
   - Validate input
   EOF
   ```

2. **Run sync (for Antigravity only):**

   ```bash
   # Cursor, Claude, Gemini see changes immediately via symlinks
   # Antigravity needs sync to copy files
   ./.agents/rules/sync-rules.sh
   ```

3. **Verify propagation:**
   ```bash
   # Check all platforms
   ls .cursor/rules/security.md
   ls .claude/rules/security.md
   ls .gemini/rules/security.md
   ls .agent/rules/security.md
   ```

### Adding New Skills

Skills use the same synchronization approach:

1. **Create skill in source:**

   ```bash
   mkdir -p .agents/skills/my-skill
   cat > .agents/skills/my-skill/SKILL.md << 'EOF'
   ---
   name: my-skill
   description: My custom skill
   ---

   # My Skill

   Skill content here.
   EOF
   ```

2. **Changes propagate automatically:**

   ```bash
   # Symlinks mean instant propagation for Cursor, Claude, Gemini
   ls .cursor/skills/my-skill
   ls .claude/skills/my-skill
   ls .gemini/skills/my-skill
   ```

3. **For Antigravity (if needed):**
   ```bash
   # Antigravity uses selective symlinks
   # May need manual setup - see .agent/skills/
   ```

## Verification

### Verify Symlinks

**Check symlink targets:**

```bash
readlink .cursor/rules    # Should output: ../.agents/rules
readlink .cursor/skills   # Should output: ../.agents/skills
readlink .claude/rules    # Should output: ../.agents/rules
readlink .claude/skills   # Should output: ../.agents/skills
readlink .gemini/rules    # Should output: ../.agents/rules
readlink .gemini/skills   # Should output: ../.agents/skills
```

**Visual verification:**

```bash
ls -la .cursor/rules .cursor/skills
# Output should show symlinks:
# lrwxr-xr-x ... rules -> ../.agents/rules
# lrwxr-xr-x ... skills -> ../.agents/skills
```

### Verify File Access

**Test reading files through symlinks:**

```bash
# Read rule through Cursor symlink
cat .cursor/rules/core-principles.md

# List skills through Claude symlink
ls .claude/skills/

# Check Gemini rules
ls .gemini/rules/*.md

# Check Antigravity (copied)
ls .agent/rules/*.md
```

### Verify in Agents

**Cursor:**

1. Open Cursor
2. Check Settings ‚Üí MCP/Skills
3. Verify skills appear

**Claude Code:**

```bash
# List installed skills
claude mcp list

# Try using a skill
claude /find-skills
```

**Gemini CLI:**

```bash
# List MCP servers and skills
gemini mcp list

# Try using a skill
gemini /find-skills
```

**Antigravity:**

```bash
# Verify rules copied
ls ~/.gemini/antigravity/.agent/rules/

# Verify skills accessible
ls ~/.gemini/antigravity/.agent/skills/
```

## Troubleshooting

### Symlinks Not Created

**Symptoms:** Directories instead of symlinks

**Diagnosis:**

```bash
ls -la .cursor/rules
# If it shows 'd' instead of 'l', it's a directory not symlink
```

**Solution:**

```bash
# Re-run sync script
./.agents/rules/sync-rules.sh

# Or manually create
rm -rf .cursor/rules
ln -s ../.agents/rules .cursor/rules
```

### Symlinks Point to Wrong Location

**Symptoms:** Broken symlinks or wrong targets

**Diagnosis:**

```bash
readlink .cursor/rules
# Should output: ../.agents/rules
```

**Solution:**

```bash
# Remove and recreate
rm .cursor/rules
ln -s ../.agents/rules .cursor/rules

# Or re-run sync
./.agents/rules/sync-rules.sh
```

### Changes Not Propagating

**For Cursor/Claude/Gemini:**

```bash
# Verify symlink exists
ls -la .cursor/rules

# Verify source file exists
ls .agents/rules/core-principles.md

# Verify symlink target is correct
readlink .cursor/rules
```

**For Antigravity:**

```bash
# Antigravity uses copies, not symlinks
# Re-run sync to copy updated files
./.agents/rules/sync-rules.sh

# Verify files copied
ls .agent/rules/
```

### Script Fails with Errors

**Missing source directory:**

```
‚ùå Rules source directory not found: .agents/rules
```

**Solution:**

```bash
# Check if source exists
ls -la .agents/rules
ls -la .agents/skills

# Clone may be incomplete - re-clone or create directories
git clone <repo-url>
```

**Permission denied:**

```
‚ùå Failed to create symlink
```

**Solution:**

```bash
# Check permissions
ls -la .cursor/

# Make script executable
chmod +x .agents/rules/sync-rules.sh

# Run with correct permissions
./.agents/rules/sync-rules.sh
```

### Antigravity-Specific Issues

**MCP servers not working:**

Antigravity does NOT support project-level MCP configurations. Must be configured globally.

**Solution:**

```bash
# Edit global config
vim ~/.gemini/antigravity/mcp_config.json

# See guide
cat docs/guides/mcp/ANTIGRAVITY_SETUP.md
```

**Rules not updated:**

Antigravity uses copied files, not symlinks.

**Solution:**

```bash
# Re-run sync to copy updated rules
./.agents/rules/sync-rules.sh

# Verify copied
ls .agent/rules/
```

## Advanced Usage

### Re-sync After Git Pull

After pulling changes that update rules/skills:

```bash
# For Cursor/Claude/Gemini - no action needed (symlinks)
# For Antigravity - re-sync to copy new rules
./.agents/rules/sync-rules.sh
```

### Clean Slate Re-sync

To completely rebuild synchronization:

```bash
# Remove all agent directories
rm -rf .cursor/rules .cursor/skills
rm -rf .claude/rules .claude/skills
rm -rf .gemini/rules
rm -rf .agent/rules

# Re-run sync
./.agents/rules/sync-rules.sh
```

### Custom Sync for Single Platform

To sync only one platform:

```bash
# Edit sync-rules.sh to comment out platforms
# Or manually create symlinks:

# Cursor only
ln -s ../.agents/rules .cursor/rules
ln -s ../.agents/skills .cursor/skills
```

## Integration with Git

### What Gets Committed

**Committed:**

- `.agents/rules/*.md` - Source rules
- `.agents/skills/` - Source skills
- `.agents/rules/sync-rules.sh` - Sync script
- `.cursor/rules`, `.claude/rules`, etc. - Symlinks themselves
- `.agent/rules/*.md` - Copied files for Antigravity

**Not committed:**

- Symlink contents (Git stores symlinks, not contents)
- Temporary files
- Personal IDE settings

### Cloning Behavior

When cloning the repository:

```bash
git clone <repo-url>
cd <repo>

# Symlinks are automatically restored
ls -la .cursor/rules  # Shows symlink

# Antigravity rules need sync (first time)
./.agents/rules/sync-rules.sh
```

## Comparison with MCP Sync

| Feature       | MCP Sync                            | Rules Sync                              |
| ------------- | ----------------------------------- | --------------------------------------- |
| Script        | `.agents/mcp/sync-mcp.sh`           | `.agents/rules/sync-rules.sh`           |
| Source        | `.agents/mcp/mcp-servers.json`      | `.agents/rules/*.md`, `.agents/skills/` |
| Method        | Generate configs                    | Create symlinks + copy                  |
| Platforms     | Cursor, Claude, Gemini, Antigravity | Same                                    |
| Run Frequency | After editing source JSON           | After adding rules (Antigravity only)   |
| Idempotent    | ‚úÖ Yes                              | ‚úÖ Yes                                  |

## Best Practices

### Do's

‚úÖ **Run sync after adding rules** (for Antigravity)
‚úÖ **Use dry-run mode** before actual sync
‚úÖ **Verify symlinks** after sync
‚úÖ **Edit source files only** (in `.agents/`)
‚úÖ **Commit source and symlinks** to Git
‚úÖ **Test in all agents** after changes

### Don'ts

‚ùå **Don't edit files through symlinks** - edit source instead
‚ùå **Don't delete .agents/ directory** - it's the source of truth
‚ùå **Don't manually create rules in agent directories** - use source
‚ùå **Don't commit broken symlinks** - verify before committing
‚ùå **Don't skip Antigravity sync** - it needs copies

## Related Documentation

- **Core Principles:** [.agents/rules/core-principles.md](../../../.agents/rules/core-principles.md)
- **Code Style:** [.agents/rules/code-style.md](../../../.agents/rules/code-style.md)
- **MCP Sync Guide:** [../mcp/mcp-setup-guide.md](../mcp/mcp-setup-guide.md)
- **Antigravity Limitations:** [../mcp/ANTIGRAVITY_LIMITATION.md](../mcp/ANTIGRAVITY_LIMITATION.md)
- **Skills Management:** [../../guidelines/team-conventions/skills-management-guidelines.md](../../guidelines/team-conventions/skills-management-guidelines.md)

## References

- **Sync Script:** `.agents/rules/sync-rules.sh`
- **MCP Sync Script:** `.agents/mcp/sync-mcp.sh` (similar pattern)
- **Project README:** `README.md`
- **Claude.md:** `.claude/CLAUDE.md`

## Support

**Issues:**

- Check [Troubleshooting](#troubleshooting) section above
- Review related documentation
- Verify source directories exist
- Try dry-run mode first

**Questions:**

- See `.agents/rules/core-principles.md` for architecture
- See `docs/references/rules/` for technical details
- See `docs/guidelines/team-conventions/skills-management-guidelines.md` for skills
